add destructors to Hasher

Please add virtual destructor to Hasher and Destructors to all the subclass. Delete function is not working properly with kDataframe

Ok, here is the output of the compilation after adding the compiler flag.

```
[ 11%] Building CXX object CMakeFiles/kmerDecoder.dir/src/KD_kmers.cpp.o
In file included from /home/mabuelanin/dib-dev/kmerDecoder/include/kmerDecoder.hpp:1,
                 from /home/mabuelanin/dib-dev/kmerDecoder/src/KD_kmers.cpp:1:
/home/mabuelanin/dib-dev/kmerDecoder/lib/kseq/include/kseq/kseq.h:42:1: warning: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++17’
   42 | inline bool delimeter_size = false;
      | ^~~~~~
[ 22%] Building CXX object CMakeFiles/kmerDecoder.dir/src/kmerDecoder.cpp.o
In file included from /home/mabuelanin/dib-dev/kmerDecoder/include/kmerDecoder.hpp:1,
                 from /home/mabuelanin/dib-dev/kmerDecoder/src/kmerDecoder.cpp:1:
/home/mabuelanin/dib-dev/kmerDecoder/lib/kseq/include/kseq/kseq.h:42:1: warning: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++17’
   42 | inline bool delimeter_size = false;
      | ^~~~~~
/home/mabuelanin/dib-dev/kmerDecoder/src/kmerDecoder.cpp:42:1: warning: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++1 
’
   42 | inline flat_hash_map<std::tuple<enum readingModes, enum hashingModes>, bool> allowed_modes ({
      | ^~~~~~
/home/mabuelanin/dib-dev/kmerDecoder/src/kmerDecoder.cpp: In static member function ‘static kmerDecoder* kmerDecoder::getInstance(readingModes, hashingModes, std::map<std::__cxx11::basic_string<char>, int>)’:
/home/mabuelanin/dib-dev/kmerDecoder/src/kmerDecoder.cpp:93:5: warning: control reaches end of non-void function [-Wreturn-type]
   93 |     }
      |     ^
[ 33%] Building CXX object CMakeFiles/kmerDecoder.dir/src/KD_skipmers.cpp.o
In file included from /home/mabuelanin/dib-dev/kmerDecoder/include/kmerDecoder.hpp:1,
                 from /home/mabuelanin/dib-dev/kmerDecoder/src/KD_skipmers.cpp:1:
/home/mabuelanin/dib-dev/kmerDecoder/lib/kseq/include/kseq/kseq.h:42:1: warning: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++17’
   42 | inline bool delimeter_size = false;
      | ^~~~~~
[ 44%] Building CXX object CMakeFiles/kmerDecoder.dir/src/KD_minimizers.cpp.o
In file included from /home/mabuelanin/dib-dev/kmerDecoder/include/kmerDecoder.hpp:1,
                 from /home/mabuelanin/dib-dev/kmerDecoder/src/KD_minimizers.cpp:13:
/home/mabuelanin/dib-dev/kmerDecoder/lib/kseq/include/kseq/kseq.h:42:1: warning: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++17’
   42 | inline bool delimeter_size = false;
      | ^~~~~~
[ 55%] Building CXX object CMakeFiles/kmerDecoder.dir/src/KD_aakmers.cpp.o
In file included from /home/mabuelanin/dib-dev/kmerDecoder/include/kmerDecoder.hpp:1,
                 from /home/mabuelanin/dib-dev/kmerDecoder/src/KD_aakmers.cpp:1:
/home/mabuelanin/dib-dev/kmerDecoder/lib/kseq/include/kseq/kseq.h:42:1: warning: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++17’
   42 | inline bool delimeter_size = false;
      | ^~~~~~
[ 66%] Building CXX object CMakeFiles/kmerDecoder.dir/src/hashUtils/aaHasher.cpp.o
[ 77%] Building CXX object CMakeFiles/kmerDecoder.dir/src/hashUtils/hashutil.cpp.o
[ 88%] Building CXX object CMakeFiles/kmerDecoder.dir/src/Utils/kmer.cpp.o
[100%] Linking CXX static library libkmerDecoder.a
[100%] Built target kmerDecoder
```

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -12,6 +12,7 @@ env:
   CTEST_OUTPUT_ON_FAILURE: 1
   CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
   CPM_SOURCE_CACHE: ${{ github.workspace }}/cpm_modules
+  CC: g++
 
 jobs:
   build:
@@ -27,8 +28,8 @@ jobs:
           path: "**/cpm_modules"
           key: ${{ github.workflow }}-cpm-modules-${{ hashFiles('**/CMakeLists.txt', '**/*.cmake') }}
 
-      - name: install boost
-        run: sudo apt-get install libboost-all-dev
+      - name: reinstall openmp
+        run:  sudo apt-get install --reinstall openmpi-bin libopenmpi-dev
         
       - uses: actions/setup-python@v2
         with:

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -42,7 +42,7 @@ jobs:
         run: cmake -Stest -Bbuild -DENABLE_TEST_COVERAGE=1 -DCMAKE_BUILD_TYPE=Debug
 
       - name: build
-        run: cmake --build build # -j4
+        run: cmake --build build -j4
 
       - name: test
         run: |

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -16,7 +16,7 @@ env:
 
 jobs:
   build:
-    runs-on: ubuntu-18.04
+    runs-on: ubuntu-latest
 
     steps:
       - uses: actions/checkout@v2

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -42,7 +42,7 @@ jobs:
         run: cmake -Stest -Bbuild -DENABLE_TEST_COVERAGE=1 -DCMAKE_BUILD_TYPE=Debug
 
       - name: build
-        run: cmake --build build -j4
+        run: cmake --build build # -j4
 
       - name: test
         run: |

--- a/include/hashUtils/aaHasher.hpp
+++ b/include/hashUtils/aaHasher.hpp
@@ -27,10 +27,13 @@ public:
     uint64_t hash(uint64_t key) override;
 
     string Ihash(uint64_t key) override;
+
+    ~aaHasher_default(){}
 };
 
 class aaHasher_dayhoff : public aaHasher_default{
 public:
     explicit aaHasher_dayhoff(uint64_t kSize);
     string Ihash(uint64_t key) override;
+    ~aaHasher_dayhoff(){}
 };

--- a/include/hashUtils/hashutil.hpp
+++ b/include/hashUtils/hashutil.hpp
@@ -50,6 +50,8 @@ public:
     }
 
     virtual Hasher *clone() { return this; };
+
+    ~Hasher(){}
 };
 
 
@@ -80,6 +82,8 @@ public:
     string Ihash(const string &key) {
         throw logic_error("Reverse Hash function is not/ cannot be implemented for this hash function.");
     }
+
+    ~bigKmerHasher(){};
 };
 
 /* ---------------------
@@ -99,6 +103,8 @@ public:
     Hasher *clone() override { return new MumurHasher(seed); }
 
     uint64_t hash(string kmer);
+
+    ~MumurHasher(){}
 };
 
 /* ---------------------
@@ -122,6 +128,8 @@ public:
     uint64_t hash(uint64_t key) override;
 
     string Ihash(uint64_t key) override;
+
+    ~IntegerHasher(){}
 };
 
 /* ---------------------
@@ -147,6 +155,8 @@ private:
     uint64_t hash(const string &key) override;
     uint64_t hash(uint64_t key) override;
     string Ihash(uint64_t key) override;
+    
+    ~noncanonical_IntegerHasher(){}
 };
 
 
@@ -172,6 +182,8 @@ public:
     uint64_t hash(uint64_t key) override;
 
     string Ihash(uint64_t key) override;
+
+    ~TwoBitsHasher(){}
 };
 
 /* ---------------------
@@ -195,6 +207,8 @@ public:
     Hasher *clone() override { return new noncanonical_TwoBitsHasher(kSize); }
 
     uint64_t hash(const string &key) override;
+
+    ~noncanonical_TwoBitsHasher(){}
 };
 
 /* ---------------------
@@ -239,6 +253,8 @@ public:
     uint64_t hash(uint64_t key) override;
 
     string Ihash(uint64_t key) override;
+
+    ~QHasher(){}
 };
 
 template<class hashFnType>
@@ -263,6 +279,7 @@ public:
         return fn(key);
     }
 
+    ~wrapperHasher(){}
 };
 
 

--- a/include/kmerDecoder.hpp
+++ b/include/kmerDecoder.hpp
@@ -163,7 +163,12 @@ public:
         return this->kSize;
     }
 
-    ~Kmers() override{}
+    ~Kmers() override{
+        delete this->hasher;
+        kseq_destroy(this->kseqObj);
+        gzclose(this->fp);
+        this->kmers.clear();
+    }
 
 };
 
@@ -244,7 +249,12 @@ public:
         return this->k;
     }
 
-    ~Skipmers(){}
+    ~Skipmers() override {
+        delete this->hasher;
+        kseq_destroy(this->kseqObj);
+        gzclose(this->fp);
+        this->kmers.clear();
+    }
 };
 
 
@@ -340,7 +350,12 @@ public:
         return this->k;
     }
 
-    ~Minimizers(){}
+    ~Minimizers(){
+        delete this->hasher;
+        kseq_destroy(this->kseqObj);
+        gzclose(this->fp);
+        this->kmers.clear();
+    }
 
 };
 
@@ -401,6 +416,11 @@ public:
         return this->kSize;
     }
 
-    ~aaKmers() override{}
+    ~aaKmers() override{
+        delete this->hasher;
+        kseq_destroy(this->kseqObj);
+        gzclose(this->fp);
+        this->kmers.clear();
+    }
 
 };

--- a/src/KD_minimizers.cpp
+++ b/src/KD_minimizers.cpp
@@ -29,7 +29,7 @@ vector<string> Minimizers::kmerize(string seq, int k)
 {
     vector<string> ret(seq.length() - k, "");
 
-#pragma omp parallel for
+// #pragma omp parallel for
     for (int i = 0; i < seq.length() - k; i++)
     {
         string s = seq.substr(i, k);

